src/openms/source/ANALYSIS/ID/SearchDatabase.cpp

```diff
@@ -0,0 +1,337 @@
+// --------------------------------------------------------------------------
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Chris Bielow $
+// $Authors: Max Alcer, Heike Einsfeld $
+// --------------------------------------------------------------------------
+
+
+#include <OpenMS/ANALYSIS/ID/SearchDatabase.h>
+
+#include <OpenMS/CHEMISTRY/AAIndex.h>
+#include <OpenMS/CHEMISTRY/AASequence.h>
+#include <OpenMS/CHEMISTRY/ModificationsDB.h>
+#include <OpenMS/CHEMISTRY/ModifiedPeptideGenerator.h>
+#include <OpenMS/CHEMISTRY/ProteaseDB.h>
+#include <OpenMS/CHEMISTRY/ProteaseDigestion.h>
+#include <OpenMS/CHEMISTRY/TheoreticalSpectrumGenerator.h>
+#include <OpenMS/CONCEPT/Constants.h>
+#include <OpenMS/DATASTRUCTURES/DefaultParamHandler.h>
+#include <OpenMS/DATASTRUCTURES/Param.h>
+#include <OpenMS/FORMAT/FASTAFile.h>
+#include <OpenMS/FORMAT/MzMLFile.h>
+#include <OpenMS/KERNEL/MSExperiment.h>
+#include <OpenMS/KERNEL/Peak1D.h>
+#include <OpenMS/MATH/MISC/MathFunctions.h>
+#include <OpenMS/QC/QCBase.h>
+
+#include <omp.h>
+#include <unordered_set>
+#include <vector>
+
+using namespace std;
+
+
+namespace OpenMS
+{
+
+using namespace Constants::UserParam;
+using namespace Math;
+
+std::vector<SearchDatabase::Peptide_> SearchDatabase::generate_peptides_(const std::vector<FASTAFile::FASTAEntry>& entries) const
+{
+  vector<SearchDatabase::Peptide_> all_peptides;
+
+  ModifiedPeptideGenerator::MapToResidueType fixed_modifications = ModifiedPeptideGenerator::getModifications(modifications_fixed_);
+  ModifiedPeptideGenerator::MapToResidueType variable_modifications = ModifiedPeptideGenerator::getModifications(modifications_variable_);
+
+  #pragma omp parallel
+  { 
+    ProteaseDigestion digestor;
+    digestor.setEnzyme(digestor_enzyme_);
+    digestor.setMissedCleavages(missed_cleavages_);
+    vector<SearchDatabase::Peptide_> all_peptides_pvt;
+    #pragma omp for nowait
+    for (size_t i = 0; i < entries.size(); i++)
+    {
+      vector<AASequence> peptides;
+      digestor.digest(AASequence::fromString(entries[i].sequence), peptides, peptide_min_length_, peptide_max_length_);
```

Comment:
I think there is a faster string_view variant

Reply:
https://github.com/OpenMS/OpenMS/blob/183363254a9b6b4f66dfa2cf0712f6133cd26bfa/src/openms/source/CHEMISTRY/EnzymaticDigestion.cpp#L429

Reply:
also returns the much more efficient indices